### PR TITLE
Add dynamic references and fix group membership issues in Microsoft Graph function

### PR DESCRIPTION
**Pull request by ytsarev**
### Summary

  - Implemented dynamic reference capabilities for all query types, allowing references to values in XR spec, status or function context
  - Fixed group membership retrieval to properly include service principals using the expand workaround
  - Corrected user type identification in group membership results
  - Simplified debug API logging for better observability
  - Enhanced reference resolution to support spec fields using spec.* syntax
  - Improved propagation of spec from observed to desired XR
  - Added example compositions that demonstrate spec reference usage
  - Updated XR example to include spec fields that can be referenced
  - Refactored code to reduce complexity and improve maintainability
  - Added comprehensive tests for spec reference functionality

### Group Membership Fixes

  - Fixed issues with group membership queries not including service principals by implementing the recommended expand workaround in
   the Microsoft Graph API
  - Corrected user type identification by properly checking properties and interfaces
  - Improved extraction of service principal and user properties from API responses
  - Simplified logging to reduce noise while maintaining essential debug information

### Dynamic Reference Capabilities

  Added support for four new reference types:
  1. groupRef: References a single group name for GroupMembership queries
  2. groupsRef: References an array of group names for GroupObjectIDs queries
  3. usersRef: References an array of user names for UserValidation queries
  4. servicePrincipalsRef: References an array of service principal names for ServicePrincipalDetails queries

  These references allow users to dynamically source values from:
  - XR status fields (e.g., status.groups)
  - Function context, including environment variables (e.g., context.[apiextensions.crossplane.io/environment].users)

### Code Improvements

  - Refactored string array resolution to use common code paths
  - Modularized reference processing to reduce cyclomatic complexity
  - Added consistent error handling and logging
  - Created example compositions showing how to use each reference type
  - Implemented comprehensive test coverage for all reference types and edge cases

### Example Usage

```yaml
  # Dynamic references in function inputs
  queryType: GroupMembership
  groupRef: status.selectedGroup

  queryType: GroupObjectIDs
  groupsRef: context.[apiextensions.crossplane.io/environment].groups

  queryType: UserValidation
  usersRef: status.users

  queryType: ServicePrincipalDetails
  servicePrincipalsRef: context.[apiextensions.crossplane.io/environment].servicePrincipalNames
```

  All these improvements make the function more flexible and maintainable while fixing critical issues with group membership retrieval.

### Testing

 ```
crossplane render xr.yaml group-membership-example.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
---
apiVersion: example.crossplane.io/v1
kind: XR
metadata:
  name: example-xr
status:
  conditions:
  - lastTransitionTime: "2024-01-01T00:00:00Z"
    reason: Available
    status: "True"
    type: Ready
  groupMembers:
  - displayName: Yury Tsarev
    id: <redacted>
    type: user
  - displayName: yury-upbound-oidc-provider
    id: <redacted>
    type: servicePrincipal
---
```
Both User and servicePrincipal group members are in the list 👍  

Dynamic `usersRef` variations:

```shell
crossplane render xr.yaml user-validation-example-status-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
```

```shell
crossplane render xr.yaml user-validation-example-context-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc --extra-resources=envconfig.yaml
```

Dynamic `groupRef` variations:

```shell
crossplane render xr.yaml group-membership-example-status-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
```

```shell
crossplane render xr.yaml group-membership-example-context-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc --extra-resources=envconfig.yaml
```

Dynamic `groupsRef` variations:

```shell
crossplane render xr.yaml group-objectids-example-status-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
```

```shell
crossplane render xr.yaml group-objectids-example-context-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc --extra-resources=envconfig.yaml
```

Dynamic `servicePrinicpalsRef` variations:

```shell
crossplane render xr.yaml service-principal-example-status-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
```

```shell
crossplane render xr.yaml service-principal-example-context-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc --extra-resources=envconfig.yaml
```

specRef tests

```shell
crossplane render xr.yaml user-validation-example-spec-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
crossplane render xr.yaml group-membership-example-spec-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
crossplane render xr.yaml group-objectids-example-spec-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
crossplane render xr.yaml service-principal-example-spec-ref.yaml functions.yaml --function-credentials=./secrets/azure-creds.yaml -rc
```

All examples are documented and working as expected.

I have:

- [X] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit tests for my change.

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

**Review by stevendborrelli — APPROVED** [review]

File: example/functions.yaml [hunk 1]
```diff
@@ -8,3 +8,11 @@ metadata:
     render.crossplane.io/runtime: Development
 spec:
   package: xpkg.upbound.io/upbound/function-msgraph:v0.1.0
+---
```
[the remark this answers is not in the document]
Comment (ytsarev) [reply example/functions.yaml:11]: thanks for the catch, amended 👍
